examples/pmp_convor.py

```diff
@@ -0,0 +1,212 @@
+                            )
+
+                            variable_names = [
+                                ("S", idx_img, idx_feat, idx_s_height, idx_s_width),
+                                (
+                                    "W",
+                                    idx_chan,
+                                    idx_feat,
+                                    idx_feat_height,
+                                    idx_feat_width,
+                                ),
+                                SW_var,
+                            ]
+
+                            fg.add_factor_by_type(
+                                variable_names=variable_names,
+                                factor_type=logical.ANDFactor,
+                            )
+
+                            X_var = (idx_img, idx_chan, idx_img_height, idx_img_width)
+                            if X_var not in variable_names_for_OR_factors:
+                                variable_names_for_OR_factors[X_var] = [SW_var]
+                            else:
+                                variable_names_for_OR_factors[X_var].append(SW_var)
+
+
+# Add ORFactors
+for X_var, variable_names_for_OR_factor in variable_names_for_OR_factors.items():
+    fg.add_factor_by_type(
+        variable_names=variable_names_for_OR_factor + [("X",) + X_var],  # type: ignore
+        factor_type=logical.ORFactor,
+    )
+
+for factor_type, factors in fg.factors.items():
+    print(f"The factor graph contains {len(factors)} {factor_type}")
+
+# %% [markdown]
+# ### Run inference and visualize results
+
+# %% [markdown]
+# PMP perturbs the model by adding Gumbel noise to unary potentials, then samples from the joint posterior *p(S,W | X)*.
+
+# %%
+run_bp, get_bp_state, get_beliefs = graph.BP(fg.bp_state, 3000)
+
+# %% [markdown]
+# We first compute the evidence without perturbation, similarly to the PMP paper
+
+# %%
+from scipy.special import logit as invsigmoid
+
+pe = 1e-100
+pS = 1e-100
+pW = 0.25
+
+uS = np.zeros((S.shape) + (2,))
+uS[..., 1] = invsigmoid(pS)
+
+uW = np.zeros((W.shape) + (2,))
+uW[..., 1] = invsigmoid(pW)
+
+uX = np.zeros((data.shape) + (2,))
+uX[..., 0] = (2 * data - 1) * invsigmoid(pe)
+
+# %% [markdown]
+# We draw a batch of samples from the posterior in parallel by transforming `run_bp`/`get_beliefs` with `jax.vmap`
+
+# %%
+np.random.seed(seed=42)
+n_samples = 4
+
+bp_arrays = jax.vmap(functools.partial(run_bp, damping=0.5), in_axes=0, out_axes=0)(
+    evidence_updates={
+        "S": uS[None] + np.random.gumbel(size=(n_samples,) + uS.shape),
+        "W": uW[None] + np.random.gumbel(size=(n_samples,) + uW.shape),
+        "SW": np.zeros(shape=(n_samples,) + SW.shape),
+        "X": uX[None] + np.zeros(shape=(n_samples,) + uX.shape),
+    },
+)
+beliefs = jax.vmap(get_beliefs, in_axes=0, out_axes=0)(bp_arrays)
+map_states = graph.decode_map_states(beliefs)
+
+# %% [markdown]
+# Visualizing the MAP decoding, we see that we have 4 samples from the posterior!
```

Comment:
More comments on what the visualization means. For example it's probably not immediately clear that each row corresponds to a sample, or why there are 5 smaller images per row, or why these smaller images are 6x6, or how these images are connected to the above visualization of the data.

Something that might also be helpful is to visualize also S of a particular image and how we combine S and W to reconstruct the original image and what those reconstructions look like when compared with original images. I'll leave it to you to decide whether to include that.